# Analisis Eksplorasi Data (EDA) Iklim dan Banjir Jakarta 2016 - 2020

**Topik:** Smart City
**Nomor Kelompok:** Kelompok 11 
**Anggota Kelompok:**
1. Darrell Theodore Ang Nitbani - 5027261048
2. Rayhana Muthia Hanin - 5027261084
3. Fakhri Abdullah - 5027261106

## Latar Belakang dan Pertanyaan Penelitian

**Latar Belakang:**  
Penelitian ini bertujuan untuk menganalisis sejauh mana pengaruh kondisi cuaca harian, seperti tingkat curah hujan, kelembapan udara, dan kecepatan angin, terhadap frekuensi kejadian banjir di wilayah DKI Jakarta. Topik ini sangat menarik untuk diteliti karena adanya disparitas spasial yang signifikan antarwilayah; kawasan Jakarta Timur dan Jakarta Selatan tercatat sebagai area yang kerap mengalami banjir, sementara Jakarta Pusat cenderung lebih jarang terdampak. Selain itu, eksplorasi awal pada dataset menunjukkan adanya karakteristik temporal yang khas di mana hari-hari kejadian banjir terpusat pada bulan-bulan tertentu (seperti bulan Februari), serta ditemukannya beberapa anomali data (seperti pencatatan suhu ekstrem). Melalui analisis eksplorasi data (EDA) ini, diharapkan pola cuaca pemicu banjir dapat teridentifikasi secara terstruktur sekaligus menjadi landasan dalam pengembangan model prediksi banjir yang lebih akurat di masa mendatang.

**Pertanyaan Penelitian:**
1. "Bagaimana perbedaan karakteristik cuaca harian (curah hujan RR, kelembapan RH_avg, dan kecepatan angin ff_x) antara Jakarta Timur & Jakarta Selatan dibandingkan dengan Jakarta Pusat, serta sejauh mana perbedaan tersebut memicu ketimpangan frekuensi kejadian banjir (flood) di wilayah-wilayah tersebut?"
2. "Bagaimana gambaran pola variabel cuaca harian (RR, RH_avg, ff_x, Tavg) pada hari kejadian banjir (flood = 1) di bulan-bulan rawan seperti Februari, dan seberapa efektif variabel cuaca harian murni tersebut dalam membedakan hari banjir dengan hari normal?"

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

df = pd.read_csv("data/data_finish.csv")

print (df.shape)
df

(6308, 15)


,date,Tn,Tx,Tavg,RH_avg,RR,ss,ff_x,ddd_x,ff_avg,ddd_car,station_id,station_name,region_name,flood
0,2016-01-01,26.0,34.8,28.6,81.0,NaN,5.8,5.0,280.0,2.0,S,96733,Stasiun Klimatologi Banten,Jakarta Selatan,0
1,2016-01-02,25.6,33.2,27.0,88.0,1.6,8.7,4.0,290.0,2.0,W,96733,Stasiun Klimatologi Banten,Jakarta Selatan,1
2,2016-01-03,24.4,34.9,28.1,80.0,33.8,5.4,4.0,280.0,2.0,SW,96733,Stasiun Klimatologi Banten,Jakarta Selatan,1
3,2016-01-04,24.8,33.6,29.2,81.0,NaN,6.6,3.0,200.0,1.0,S,96733,Stasiun Klimatologi Banten,Jakarta Selatan,0
4,2016-01-05,25.8,33.6,26.7,91.0,NaN,3.2,3.0,180.0,1.0,S,96733,Stasiun Klimatologi Banten,Jakarta Selatan,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6303,2018-12-27,23.8,32.0,28.0,70.0,NaN,NaN,12.0,180.0,5.0,W,96747,Halim Perdana Kusuma Jakarta,Jakarta Timur,0
6304,2018-12-28,24.0,33.4,28.5,69.0,NaN,NaN,14.0,250.0,3.0,SE,96747,Halim Perdana Kusuma Jakarta,Jakarta Timur,0
6305,2018-12-29,25.2,33.4,28.7,70.0,NaN,NaN,14.0,120.0,5.0,SW,96747,Halim Perdana Kusuma Jakarta,Jakarta Timur,0
6306,2018-12-30,24.0,34.4,30.0,64.0,NaN,NaN,14.0,240.0,5.0,W,96747,Halim Perdana Kusuma Jakarta,Jakarta Timur,0


Tabel di atas menampilkan 5 baris pertama dari dataset kami yang berisi data cuaca dan kejadian banjir. Setiap baris pada dataset ini mewakili satu entitas observasi/pengukuran harian.

## Sumber Data dan Kamus Data

* **Sumber Data:** Data Iklim dan Kejadian Banjir Jakarta (2016-2020)
* **Pengumpul Data:** BMKG / BPBD DKI Jakarta

### Kamus Data

| Kolom | Arti | Jenis Variabel | Skala | Satuan | Contoh |
| :--- | :--- | :--- | :--- | :--- | :--- |
| **date** | Tanggal pencatatan cuaca dan kejadian | Kuantitatif | Rasio | YYYY-MM-DD | 2020-02-01 ||
| **Tn** | Suhu minimum harian | Kuantitatif Kontinu | Rasio | °C | 24.2 |
| **Tx** | Suhu maksimum harian | Kuantitatif Kontinu | Rasio | °C | 33.0 |
| **Tavg** | Suhu rata-rata harian | Kuantitatif Kontinu | Rasio | °C | 28.5 |
| **RH_avg** | Kelembapan udara rata-rata | Kuantitatif Kontinu | Rasio | % | 85.0 |
| **RR** | Curah hujan harian | Kuantitatif Kontinu | Rasio | mm | 45.2 |
| **ss** | Lamanya penyinaran matahari | Kuantitatif Kontinu | Rasio | Jam | 4.5 |
| **ff_x** | Kecepatan angin maksimum | Kuantitatif Kontinu | Rasio | m/s | 6.0 |
| **ff_avg** | Kecepatan angin rata-rata | Kuantitatif Kontinu | Rasio | m/s | 2.4 |
| **flood** | Status kejadian banjir (0 = Tidak, 1 = Banjir) | Kualitatif | Nominal / Biner | - | 1 |

In [2]:
df.info()                 # tipe data setiap kolom

<class 'pandas.DataFrame'>
RangeIndex: 6308 entries, 0 to 6307
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   date          6308 non-null   str    
 1   Tn            5996 non-null   float64
 2   Tx            6095 non-null   float64
 3   Tavg          6262 non-null   float64
 4   RH_avg        6256 non-null   float64
 5   RR            3993 non-null   float64
 6   ss            5049 non-null   float64
 7   ff_x          6286 non-null   float64
 8   ddd_x         6286 non-null   float64
 9   ff_avg        6215 non-null   float64
 10  ddd_car       6207 non-null   str    
 11  station_id    6308 non-null   int64  
 12  station_name  6308 non-null   str    
 13  region_name   6308 non-null   str    
 14  flood         6308 non-null   int64  
dtypes: float64(9), int64(2), str(4)
memory usage: 739.3 KB


In [3]:
df.isna().sum()           # jumlah data kosong per kolom

date               0
Tn               312
Tx               213
Tavg              46
RH_avg            52
RR              2315
ss              1259
ff_x              22
ddd_x             22
ff_avg            93
ddd_car          101
station_id         0
station_name       0
region_name        0
flood              0
dtype: int64

In [4]:
df.duplicated().sum()     # jumlah baris duplikat

np.int64(0)

### Penjelasan Kualitas & Pembersihan Data

Berdasarkan pemeriksaan data, berikut temuan dan keputusan penanganannya:

1. **Penghapusan Fitur yang Tidak Relevan:**
   * **Temuan:** Kolom `station_id` (kode ID angka), `station_name` dan `ddd_x` bersifat redundan. Informasi lokasi dan arah angin sudah diwakili secara lebih jelas oleh kolom `station_name` dan `ddd_x`.
   * **Keputusan:** **Dihapus (Dropped)** dari dataset.
   * **Alasan:** Menyederhanakan dataset agar fokus pada variabel cuaca numerik dan mengurangi beban pemrosesan data.

2. **Data Kosong (*Missing Values*):**
   * **Temuan:** Kolom `RR` (Curah Hujan) memiliki data kosong paling banyak (sekitar 36,7%).
   * **Keputusan:** **Dibiarkan (tidak dihapus)**.
   * **Alasan:** Menghapus baris kosong pada `RR` akan membuang sepertiga data utama. Data dibiarkan utuh untuk tahap eksplorasi awal.

In [7]:
dfclean = df.drop(columns=['station_id', 'station_name', 'ddd_x'], errors='ignore')
dfclean['date'] = pd.to_datetime(dfclean['date'])
dfclean['RR'] = dfclean['RR'].fillna(0)

num_cols = ['Tn', 'Tx', 'Tavg', 'RH_avg', 'ss', 'ff_x', 'ff_avg']
for col in num_cols:
    dfclean[col] = dfclean.groupby('region_name')[col].transform(lambda x: x.fillna(x.median()))
    dfclean[col] = dfclean[col].fillna(dfclean[col].median())

if 'ddd_car' in dfclean.columns:
    global_mode_ddd = dfclean['ddd_car'].mode().iloc[0]
    dfclean['ddd_car'] = dfclean['ddd_car'].fillna(
        dfclean.groupby('region_name')['ddd_car'].transform(
            lambda x: x.mode().iloc[0] if not x.mode().empty else global_mode_ddd
        )
    )

outlier_cols = ['Tavg', 'RH_avg', 'ff_x', 'Tn', 'Tx']

def iqr(g, col):
    Q1, Q3 = g[col].quantile(0.25), g[col].quantile(0.75)
    IQR = Q3 - Q1
    lower, upper = Q1 - 1.5 * IQR, Q3 + 1.5 * IQR
    return g[col].clip(lower=lower, upper=upper)

for col in outlier_cols:
    dfclean[col] = dfclean.groupby('region_name', group_keys=False).apply(lambda g: iqr(g, col))

print("=== Cek Sisa Data Kosong ===")
print(dfclean.isna().sum())

print("\n=== Tipe Data Terbaru ===")
print(dfclean[['date', 'RR', 'RH_avg', 'ff_x', 'Tavg', 'flood']].dtypes)

=== Cek Sisa Data Kosong ===
date           0
Tn             0
Tx             0
Tavg           0
RH_avg         0
RR             0
ss             0
ff_x           0
ff_avg         0
ddd_car        0
region_name    0
flood          0
dtype: int64

=== Tipe Data Terbaru ===
date      datetime64[us]
RR               float64
RH_avg           float64
ff_x             float64
Tavg             float64
flood              int64
dtype: object


### Penjelasan Kualitas Data & Penanganan Pembersihan Data

1. **Konversi Tipe Data `date`:**
   * **Tindakan:** Mengubah tipe data kolom `date` dari *string* menjadi `datetime64`.
   * **Alasan:** Perubahan ini diperlukan secara mendasar untuk menjawab **Pertanyaan #2**, yaitu mengekstrak komponen bulan agar dapat menganalisis tren banjir dan karakteristik cuaca harian khusus pada bulan-bulan rawan seperti bulan Februari.

2. **Pengisian Data Kosong (*Imputation*):**
   * **Curah Hujan (`RR`):** Diisi dengan nilai **0**. Dalam data meteorologi harian, hari tanpa presipitasi sering kali tidak mencatat angka sama sekali sehingga terbaca sebagai *missing/NaN*.
   * **Variabel Cuaca Numerik (`RH_avg`, `ff_x`, `Tavg`, dll.):** Diisi menggunakan nilai **Median spesifik per wilayah (`region_name`)**. Langkah ini diambil untuk menjawab **Pertanyaan #1**, di mana dinamika cuaca mikro antara Jakarta Timur, Jakarta Selatan, dan Jakarta Pusat tetap terjaga akurasinya tanpa terpengaruh rata-rata global.
   * **Arah Angin (`ddd_car`):** Diisi dengan nilai **Modus (kategori paling sering muncul)**.

3. **Penanganan Nilai Ekstrem (*Outlier*):**
   * **Tindakan:** Menerapkan teknik *IQR Capping* (pembatasan nilai ekstrem pada rentang $Q_1 - 1.5 \times IQR$ hingga $Q_3 + 1.5 \times IQR$) khusus untuk variabel `Tavg`, `RH_avg`, `ff_x`, `Tn`, dan `Tx`.
   * **Pengecualian `RR` (Curah Hujan):** Nilai *outlier* pada kolom `RR` **dibiarkan apa adanya (tidak dipotong/dihapus)**. Hal ini sengaja dilakukan untuk menjawab **Pertanyaan #3**, karena angka curah hujan yang sangat tinggi (*outlier*) merupakan sinyal cuaca ekstrem yang menjadi pemicu utama kejadian banjir (`flood = 1`).